test/util/framework/per_test_framework.go

```diff
@@ -0,0 +1,273 @@
+	"time"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/resources/armresources"
+	"github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/resources/armsubscriptions"
+	"github.com/onsi/ginkgo/v2"
+	"golang.org/x/sync/errgroup"
+	"k8s.io/apimachinery/pkg/util/rand"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+
+	hcpapi20240610 "github.com/Azure/ARO-HCP/internal/api/v20240610preview/generated"
+)
+
+type perItOrDescribeTestContext struct {
+	perBinaryInvocationTestContext *perBinaryInvocationTestContext
+
+	contextLock                   sync.RWMutex
+	knownResourceGroups           []string
+	subscriptionID                string
+	clientFactory20240610         *hcpapi20240610.ClientFactory
+	armResourcesClientFactory     *armresources.ClientFactory
+	armSubscriptionsClientFactory *armsubscriptions.ClientFactory
+}
+
+func NewInvocationContext() *perItOrDescribeTestContext {
+	tc := &perItOrDescribeTestContext{
+		perBinaryInvocationTestContext: invocationContext(),
+	}
+
+	// this construct allows us to be called before or after the test has started and still registry proper cleanup.
+	if !ginkgo.GetSuite().InRunPhase() {
+		// The order is important here: if the extension calls ginkgo.BeforeEach
+		// itself, then it can be sure that f.BeforeEach already ran when its
+		// own callback gets invoked.
+		ginkgo.BeforeEach(tc.BeforeEach, AnnotatedLocation("set up test context"))
+	} else {
+		tc.BeforeEach(context.TODO())
+	}
+
+	return tc
+}
+
+// BeforeEach gives a chance for initialization (none yet) and registers the cleanup
+func (tc *perItOrDescribeTestContext) BeforeEach(ctx context.Context) {
+	// DeferCleanup, in contrast to AfterEach, triggers execution in
+	// first-in-last-out order. This ensures that the framework instance
+	// remains valid as long as possible.
+	//
+	// In addition, AfterEach will not be called if a test never gets here.
+	ginkgo.DeferCleanup(tc.deleteCreatedResources, AnnotatedLocation("tear down test context"), ginkgo.NodeTimeout(45*time.Minute))
+
+	// Registered later and thus runs before deleting namespaces.
+	ginkgo.DeferCleanup(tc.collectDebugInfo, AnnotatedLocation("dump debug info"), ginkgo.NodeTimeout(45*time.Minute))
+}
+
+// deleteCreatedResources deletes what was created that we know of.
+func (tc *perItOrDescribeTestContext) deleteCreatedResources(ctx context.Context) {
+	tc.contextLock.RLock()
+	defer tc.contextLock.RUnlock()
+
+	// deletion takes a while, it's worth it to do this in parallel
+	waitGroup, ctx := errgroup.WithContext(ctx)
+	for _, resourceGroupName := range tc.knownResourceGroups {
+		waitGroup.Go(func() error {
+			// prevent a stray panic from exiting the process. Don't do this generally because ginkgo/gomega rely on panics ot funcion.
+			utilruntime.HandleCrashWithContext(ctx)
+
+			return tc.cleanupResourceGroup(ctx, resourceGroupName)
+		})
+	}
+	if err := waitGroup.Wait(); err != nil {
+		// remember that Wait only shows the first error, not all the errors.
```

Comment:
That's because it doesn't actually collect all the errors, waitgroup crashes out the moment any one `.Go()` call exits

Reply:
> That's because it doesn't actually collect all the errors, waitgroup crashes out the moment any one `.Go()` call exits

Yes.  I'm ok with this, but I wanted future generations to see without digging into doc.